File: LeetCode_rename/U_300_LongestIncreasingSubsequence.py
# https://leetcode.com/problems/longest-increasing-subsequence
# 2022 2.22

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        
        # v1, copied dp, slow, 5600ms
        dp=[1]*len(nums)
        for i in range(0,len(nums)):
            for j in range(0,i-1):
                if nums[i]>nums[j]:
                    dp[i]=max(dp[i],dp[j]+1)
            # print(dp)
        return max(dp)

File: LeetCode/U_300_LongestIncreasingSubsequence.java
// https://leetcode.com/problems/longest-increasing-subsequence
// 2022年11月09日 18:01:38

class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] dp = new int[nums.length]; // recort max subsequence length till now
        Arrays.fill(dp, 1);
        int ret = 1;
        for (int i = 1; i < nums.length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            ret = Math.max(ret, dp[i]);
        }
        return ret;
    }
}